### ISSUE TITLE
Race condition when creating Guice Modules

### ISSUE THREAD

**Issue by lucianoRM** [testng-team/testng#3050]
### TestNG Version
7.7.1

### Expected behavior
Each module should be instanciated once for the duration of the Suite execution

### Actual behavior
When running multiple tests concurrently, we found an scenario where the same Guice Module is created more than once. Leading to unexpected errors.

The problem I suspect is happening is that RetryAnalyzers are not instantiated until you have a failure in some test. Then, if multiple tests fail at the same time, they all want to create the RetryAnalyzer together. If instantiation depends on Guice Inyection, the module too needs to be instantiated, and there should be only one module available. 
However, the caching mechanism in `org.testng.internal.objects.GuiceHelper#getModules` is not synchronized. So you can have a race condition there.

### Test case sample
To test, copy all the classes under the package `some.test` and run suite `evidence.xml`. You will see that the inyected UUID is not the same for all results. When it should be since the Guice Module is marked as @ Singleton. 
You will randomly see the following error too:

```
com.google.inject.CreationException: Unable to create injector, see the following errors:

1) [Guice/BindingAlreadySet]: UUID was bound multiple times.

Bound at:
1  : EvidenceModule.randomUUID(EvidenceModule.java:9)
2  : EvidenceModule.randomUUID(EvidenceModule.java:9)

Learn more:
  https://github.com/google/guice/wiki/BINDING_ALREADY_SET

1 error

======================
Full classname legend:
======================
EvidenceModule: "some.test.EvidenceModule"
========================
End of classname legend:
========================
```

#### Guice Module
```
package some.test;

import com.google.inject.AbstractModule;
import com.google.inject.Provides;
import com.google.inject.Singleton;
import java.util.UUID;


public class EvidenceModule extends AbstractModule {

    @Provides
    @Singleton
    private UUID randomUUID() {
        return UUID.randomUUID();
    }

}
```

#### Retry Analyzer
```
package some.test;

import com.google.inject.Inject;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;
import org.testng.IRetryAnalyzer;
import org.testng.ITestResult;
import org.testng.annotations.Guice;


@Guice(modules = EvidenceModule.class)
public class EvidenceRetryAnalyzer implements IRetryAnalyzer {

    @Inject
    private UUID injectedUUID;

    @Override
    public boolean retry(ITestResult result) {
        System.out.println("Injected UUID is: " + injectedUUID);
        return false;
    }
}
```

#### TestOne
```
package some.test;

import static org.testng.AssertJUnit.fail;

import org.testng.ITestContext;
import org.testng.ITestNGMethod;
import org.testng.annotations.BeforeSuite;
import org.testng.annotations.Test;


@Test
public class EvidenceTestOne {

    @BeforeSuite(alwaysRun = true)
    public void suiteSetup(ITestContext context) {
        for (ITestNGMethod method : context.getAllTestMethods()) {
            method.setRetryAnalyzerClass(EvidenceRetryAnalyzer.class);
        }
    }

    @Test
    public void testOne() {
        fail();
    }

    @Test
    public void testTwo() {
        fail();
    }

}
```

#### Test Two
```
package some.test;

import static org.testng.AssertJUnit.fail;

import org.testng.ITestContext;
import org.testng.ITestNGMethod;
import org.testng.annotations.BeforeSuite;
import org.testng.annotations.Test;


@Test
public class EvidenceTestTwo {

    @BeforeSuite(alwaysRun = true)
    public void suiteSetup(ITestContext context) {
        for (ITestNGMethod method : context.getAllTestMethods()) {
            method.setRetryAnalyzerClass(EvidenceRetryAnalyzer.class);
        }
    }

    @Test
    public void testOne() {
        fail();
    }

    @Test
    public void testTwo() {
        fail();
    }

}
```

#### Test Three
```
package some.test;

import static org.testng.AssertJUnit.fail;

import org.testng.ITestContext;
import org.testng.ITestNGMethod;
import org.testng.annotations.BeforeSuite;
import org.testng.annotations.Test;


@Test
public class EvidenceTestThree {

    @BeforeSuite(alwaysRun = true)
    public void suiteSetup(ITestContext context) {
        for (ITestNGMethod method : context.getAllTestMethods()) {
            method.setRetryAnalyzerClass(EvidenceRetryAnalyzer.class);
        }
    }

    @Test
    public void testOne() {
        fail();
    }

    @Test
    public void testTwo() {
        fail();
    }

}
```

#### Suite config

```
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="Evidence" verbose="10" parallel="classes" thread-count="7">
    <test name="evidence">
        <packages>
            <package name="some.test.*"/>
        </packages>
    </test>
</suite>

```

### Contribution guidelines

Incase you plan to raise a pull request to fix this issue, please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

**Comment by krmahadevan** [comment]
@lucianoRM - The latest released version of TestNG is `7.9.0`. Please upgrade and try again. I don't see that issue in the latest release.

PS: You need to get rid of the `ITestContext` dependency in your `@BeforeSuite`. It's not supported.

```java
    @BeforeSuite(alwaysRun = true)
    public void suiteSetup() {
        ITestContext context = Reporter.getCurrentTestResult().getTestContext();
        for (ITestNGMethod method : context.getAllTestMethods()) {
            method.setRetryAnalyzerClass(EvidenceRetryAnalyzer.class);
        }
    }

```

Above is a sample.

**Comment by lucianoRM** [comment]
I just upgraded to 7.9.0 and can still see the issue. 

After running the suite I see in the console:

```
Injected UUID is: 449db11c-5549-4e0f-a322-df0c2fc7d46c
Injected UUID is: 94b42ada-336c-4dea-9d21-5c4cf1a9f502
Injected UUID is: 958e57bd-8e80-43ca-ba77-ec8c131a0ac2
```

This is printed by the EvidenceRetryAnalyzer, which, even if multiple instances are created, they all should receive the same UUID injected, since there should only be one EvidenceModule with a Singleton value.
